src/db/spanner/models.rs

```diff
@@ -941,55 +944,108 @@ impl SpannerDb {
 
         let mut sqltypes = HashMap::new();
 
-        if let Some(older) = older {
-            query = format!("{} AND modified < @older", query);
-            sqlparams.insert("older".to_string(), as_value(older.as_rfc3339()?));
-            sqltypes.insert("older".to_string(), as_type(TypeCode::TIMESTAMP));
-        }
-        if let Some(newer) = newer {
-            query = format!("{} AND modified > @newer", query);
-            sqlparams.insert("newer".to_string(), as_value(newer.as_rfc3339()?));
-            sqltypes.insert("newer".to_string(), as_type(TypeCode::TIMESTAMP));
-        }
-
         if !ids.is_empty() {
             query = format!("{} AND bso_id IN UNNEST(@ids)", query);
             sqlparams.insert("ids".to_owned(), as_list_value(ids.into_iter()));
         }
 
+        if let Some(timestamp) = offset.clone().unwrap_or_default().timestamp {
+            query = match sort {
+                Sorting::Newest => {
+                    sqlparams.insert("older_eq".to_string(), as_value(timestamp.as_rfc3339()?));
+                    sqltypes.insert("older_eq".to_string(), as_type(TypeCode::TIMESTAMP));
+                    format!("{} AND modified <= @older_eq", query)
+                }
+                Sorting::Oldest => {
+                    sqlparams.insert("newer_eq".to_string(), as_value(timestamp.as_rfc3339()?));
+                    sqltypes.insert("newer_eq".to_string(), as_type(TypeCode::TIMESTAMP));
+                    format!("{} AND modified >= @newer_eq", query)
+                }
+                _ => query,
+            };
+        } else {
+            if let Some(older) = older {
```

Comment:
One last thing: I'm wondering if we should just continue emitting `newer`/`older` when `newer_eq`/`older_eq` are, to match the python code.

Maybe it doesn't make sense for both clauses to be enabled, but if so maybe the validation layer should be enforcing it (`newer`/`older` parameters from the client would be disallowed when an `offset w/ timestamp` were specified). Maybe @rfk can chime in on this?

Reply:
Sure, I think it makes sense to match the python code as much as possible even if the clauses are redundant.

Reply:
I think it's reasonable for the client to continue sending `newer` even when it's using an `offset` token, because the offset token should be opaque to the client and may not encode all the info needed to reproduce the original query. As a client, I would expect to have to just append the given `offset` to the existing URL without modifying other query parameters.

Does that answer the concern? I'm not sure I fully understand what you're getting at here :-)

Reply:
Thanks @rfk, that makes sense and we just ran into a test that confirms this.